gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1360 @@
+ * verification of domain name ownership. Any request for creating a managed zone will be approved.
+ * The mock does not track quota and will allow the user to exceed it. The mock provides only basic
+ * validation of the DNS data for records of type A and AAAA. It does not validate any other record
+ * types.
+ */
+public class LocalDnsHelper {
+
+  private final ConcurrentSkipListMap<String, ProjectContainer> projects
+      = new ConcurrentSkipListMap<>();
+  private static final URI BASE_CONTEXT;
+  private static final Logger log = Logger.getLogger(LocalDnsHelper.class.getName());
+  private static final JsonFactory jsonFactory = new JacksonFactory();
+  private static final Random ID_GENERATOR = new Random();
+  private static final String VERSION = "v1";
+  private static final String CONTEXT = "/dns/" + VERSION + "/projects";
+  private static final Set<String> SUPPORTED_COMPRESSION_ENCODINGS =
+      ImmutableSet.of("gzip", "x-gzip");
+  private static final List<String> TYPES = ImmutableList.of("A", "AAAA", "CNAME", "MX", "NAPTR",
+      "NS", "PTR", "SOA", "SPF", "SRV", "TXT");
+  private static final TreeSet<String> FORBIDDEN = Sets.newTreeSet(
+      ImmutableList.of("google.com.", "com.", "example.com.", "net.", "org."));
+
+  static {
+    try {
+      BASE_CONTEXT = new URI(CONTEXT);
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(
+          "Could not initialize LocalDnsHelper due to URISyntaxException.", e);
+    }
+  }
+
+  private long delayChange;
+  private final HttpServer server;
+  private final int port;
+
+  /**
+   * For matching URLs to operations.
+   */
+  private enum CallRegex {
+    CHANGE_CREATE("POST", CONTEXT + "/[^/]+/managedZones/[^/]+/changes"),
+    CHANGE_GET("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/changes/[^/]+"),
+    CHANGE_LIST("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/changes"),
+    ZONE_CREATE("POST", CONTEXT + "/[^/]+/managedZones"),
+    ZONE_DELETE("DELETE", CONTEXT + "/[^/]+/managedZones/[^/]+"),
+    ZONE_GET("GET", CONTEXT + "/[^/]+/managedZones/[^/]+"),
+    ZONE_LIST("GET", CONTEXT + "/[^/]+/managedZones"),
+    PROJECT_GET("GET", CONTEXT + "/[^/]+"),
+    RECORD_LIST("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/rrsets");
+
+    private String method;
+    private String pathRegex;
+
+    CallRegex(String method, String pathRegex) {
+      this.pathRegex = pathRegex;
+      this.method = method;
+    }
+  }
+
+  /**
+   * Wraps DNS data by adding a timestamp and id which is used for paging and listing.
+   */
+  static class RrsetWrapper {
+    static final Function<ResourceRecordSet, RrsetWrapper> WRAP_FUNCTION =
+        new Function<ResourceRecordSet, RrsetWrapper>() {
+          @Override
+          public RrsetWrapper apply(ResourceRecordSet input) {
+            return new RrsetWrapper(input);
+          }
+        };
+    private final ResourceRecordSet rrset;
+    private final Long timestamp = System.currentTimeMillis();
```

Comment:
No, it is in fact not used. I was intended to be used before, but then we chose another solution.

I like removing the wrapper. It does not even need to be a concurrent map as we are handling it via atomic reference (and that must be preserved for transnactional behaviour of changes). I am refactoring it.